### ISSUE TITLE
Update the FPAnalyzer tool to work with the current version of Odyssey

### ISSUE THREAD

**Issue by elmisback** [elmisback/magic-markup#6]
The old tool lives at https://github.com/elmisback/magic-markup/blob/main/codetations-react/src/applications/src/FPAnalyzer.tsx, we'll want to move that over to the vscode-extension dir

**Comment by elmisback** [comment]
Existing tools.tsx file: https://github.com/elmisback/magic-markup/blob/main/vscode-extension/webview-ui/src/tools.tsx

**Comment by elmisback** [comment]
General plan:
  * [ ] Add the FPAnalyzer to tools.tsx
  * [ ] Update the language model that is called--call the LM API of the VSCode host using message passing. Add this call as a library function given to the `utils` prop for tools (like `get/setText`)
  * [ ] (optional) Give the FPAnalyzer a "human-readable expression" input field where the user can edit and save the spec after the LM suggests it based on the code
  * [ ] Update Odyssey to take the resolved expression as a URL parameter
  * [ ] Plan a 10-second clip showing the full workflow with an annotation on some Python code

**Comment by jaelafield** [comment]
Updated plan by Edward:
- [ ] Create a mock up of Odyssey in Codetations to include in the paper
        - [ ] review current paper state to inform that mock up
- [ ] Potentially add to/revise the relevant portion of the paper 

(Sick this week, so haven't made progress, but thought I'd at least record the new steps. Hopefully I got that right, we just talked about it briefly.)

**Comment by elmisback** [comment]
This looks good--there might not be enough time to get things into the paper though, if we're being realistic. I'm not going to have the corrected draft until tomorrow at this rate.

**Comment by jaelafield** [comment]
This should be moved to To Do, or maybe closed as not planned if it's indefinitely on hold/will be approached differently. (I don't seem to have permissions to do it myself.)

Apologies that I wasn't able to help with this Edward! I do think I could've cranked out a pretty solid mock up, if I had prioritized it right after you suggested that, but I didn't realize that covid & just tight paper timing would make that difficult.

**Comment by elmisback** [comment]
No worries! Things got pretty hectic. It would be nice to have something running for this before my generals on the 29th, but I will take it off assigned tasks until I have time to look at it myself.

**Comment by jaelafield** [comment]
On the Odyssey side, I finished enabling loading an expression via a url. I think this is enough to move to the Codetations side and get a basic version of the full connection working next. 

Worth noting that there is **no spec validation** set up right now. It shouldn't be hard to add (🤞 just stealing it from the main spec component in Odyssey), but I wanted to move on and not think about it yet.

**Comment by elmisback** [comment]
Could you give an example of a URL that works?

**Comment by elmisback** [comment]
Nvm, I tried the first thing I thought of and https://herbie-fp.github.io/odyssey/?spec=x*x works as expected. We should also think about allowing other expressions to be shown--basically, it would be nice if we have a bit of state to preserve on the codetations side, so this is an example of a stateful component

**Comment by jaelafield** [comment]
Oh good point! The query key it's looking for is `spec` and you need to make sure to encode tricky characters (like + is %2B)

https://herbie-fp.github.io/odyssey/?spec=x%2B1-1

**Comment by jaelafield** [comment]
When you say "preserve on the codetations side" do you mean in codetations it would have more than just "x+1-1" stored, it would also have a few alternatives and be able to pass those in?

That's definitely possible, and would be relatively quick, I think. Though (as we've talked about a lot) we're limited by what we can put in a URL, so it would require all new herbie calls on load (as this does for the single spec). 
Regardless though, I just wanted to just get the minimum viable thing from both sides going first.

Let's talk about this in the RA meeting some!

**Comment by elmisback** [comment]
Yeah, I think having alternatives would be good for helping users think about a workflow around this tool, and for justifying our technique in the paper (if you don't need state, you mostly don't need to preserve a notion of the "same abstract thing" across different versions of the code).

Notes from my plan for this feature are here:
* okay, so we can put an expression in the url
* would be nice to get some state in there, so we have to deal with state saving and can point that out to the user
* missing pieces:
    * convert code to mathjs using vscode LM api (this was working before, i think) -- requires exposing vscode LM api to extensions
    * get results/state from odyssey?
    * convert mathjs back to code for export?

**Comment by jaelafield** [comment]
I added a really basic component that links to odyssey and loads the selected text as the spec. It's just an <a> at the moment. I tried for longer than I should've probably to get it to work with a button instead, basically trying to get this onClick to work:

`onClick={vscode.env.openExternal(vscode.uri.parse(url)}
`
but getting vscode to import was causing me lots of trouble (related to the errors I showed you briefly, Edward). 
I'll probably try to get direct debugging help with this, it's not a blocker to start other parts though.

Next step is trying to get the LM hooked up for parsing text, as right now it creates the url with the annotation text as-is.

**Comment by jaelafield** [comment]
Yesterday, added alternatives to the state that can be initialized via query parameters (alternatives should be listed in a "," separated list after query param `alts`:

https://herbie-fp.github.io/odyssey/?spec=x%2B1-1&alts=x,x%2B1

Found a quick bug related to this just now, and pushed a fix.

**Comment by jaelafield** [comment]
The Odyssey annotation component now uses `lm-api-client` to ask for the text to be parsed for a math-js interpretable string which is then used to load that page in Odyssey! Needs more testing for sure to see if the prompt needs to be refined more, but it worked on all the examples I tried.

**Comment by elmisback** [comment]
Study code is here https://github.com/elmisback/codetations-study-1

**Comment by jaelafield** [comment]
Notes for next steps, paused for now besides bug fixing existing work!

In Odyssey:
- Get and store id for annotation in query params
- Access codetations server (hardcoded port) when alternatives are found (or any time alts are updated), and pass back found alternatives & id for annotation to update

In Codetations:
- Update link accessed to pass in id of annotation it's associated with
- Set up server (some localhost port open) --> may belong in annotation tracker, could be a "library" kind of server initializer which passes results back to component that accessed it to actually make updates to annotation and webview
- Component listens for that new info and stores alts in state (get annotation file from disk and update it) 
- Update link accessed in button to go to Odyssey so it opens with saved links

**Comment by jaelafield** [comment]
Haven't made progress this week yet, but I have time blocked off Thu to work on:

- a bug fix (Url used to but is no longer properly encoding "+" etc. just the new way the button re-direct forms that url I expect, haven't had a chance to mess with it) 
- testing of different languages/libraries for prompt refinement (checking out Study code and possibly using a model to generating some additional example programs to try depending on what's there)

**Comment by elmisback** [comment]
I fixed the + issue

I think the main areas to focus on are 1) adding support for inferring and
caching range settings and 2) giving users a clean workflow for taking
results back from Odyssey to their code. If a user doesn't find the
expression export tool in Odyssey due to bad discoverability, they might
still be able to use something like an auto-translator field in the tool
that handles compiling mathjs/whatever to the language it sees in the file.

**Comment by elmisback** [comment]
Closing this and opening a new issue for Odyssey.
